onmt/encoders/bert.py

```diff
@@ -0,0 +1,144 @@
+            is list else opt.attention_dropout,
+            opt.max_relative_positions)
+
+    def forward(self, input_ids, token_type_ids=None, input_mask=None,
+                output_all_encoded_layers=False):
+        """
+        Args:
+            input_ids (Tensor): ``(B, S)``, padding ids=0
+            token_type_ids (Tensor): ``(B, S)``, A(0), B(1), pad(0)
+            input_mask (Tensor): ``(B, S)``, 1 for masked (padding)
+            output_all_encoded_layers (bool): if out contain all hidden layer
+        Returns:
+            all_encoder_layers (list of Tensor): ``(B, S, H)``, token level
+            pooled_output (Tensor): ``(B, H)``, sequence level
+        """
+
+        # OpenNMT waiting for mask of size [B, 1, T],
+        # while in MultiHeadAttention part2 -> [B, 1, 1, T]
+        if input_mask is None:
+            # shape: 2D tensor [batch, seq]
+            padding_idx = self.embeddings.word_padding_idx
+            # shape: 2D tensor [batch, seq]: 1 for tokens, 0 for paddings
+            input_mask = input_ids.data.eq(padding_idx)
+        # [batch, seq] -> [batch, 1, seq]
+        attention_mask = input_mask.unsqueeze(1)
+
+        # embedding vectors: [batch, seq, hidden_size]
+        out = self.embeddings(input_ids, token_type_ids)
+        all_encoder_layers = []
+        for layer in self.encoder:
+            out = layer(out, attention_mask)
+            if output_all_encoded_layers:
+                all_encoder_layers.append(self.layer_norm(out))
+        out = self.layer_norm(out)
+        if not output_all_encoded_layers:
+            all_encoder_layers.append(out)
+        pooled_out = self.pooler(out)
+        return all_encoder_layers, pooled_out
+
+    def update_dropout(self, dropout):
+        self.dropout = dropout
+        self.embeddings.update_dropout(dropout)
+        for layer in self.encoder:
+            layer.update_dropout(dropout)
+
+
+class BertPooler(nn.Module):
+    def __init__(self, hidden_size):
+        """A pooling block (Linear layer followed by Tanh activation).
+
+        Args:
+            hidden_size (int): size of hidden layer.
+        """
+
+        super(BertPooler, self).__init__()
+        self.dense = nn.Linear(hidden_size, hidden_size)
+        self.activation_fn = nn.Tanh()
+
+    def forward(self, hidden_states):
+        """hidden_states[:, 0, :] --> {Linear, Tanh} --> Returns.
+
+        Args:
+            hidden_states (Tensor): last layer's hidden_states, ``(B, S, H)``
+        Returns:
+            pooled_output (Tensor): transformed output of last layer's hidden
+        """
+
+        first_token_tensor = hidden_states[:, 0, :]  # [batch, d_model]
+        pooled_output = self.activation_fn(self.dense(first_token_tensor))
+        return pooled_output
+
+
+class BertLayerNorm(nn.Module):
+    def __init__(self, hidden_size, eps=1e-12):
+        """Layernorm module in the TF style(epsilon inside the square root).
```

Comment:
Does PyTorch implement it differently?

Reply:
I adapt the BertLayerNorm from huggingface/pytorch-transformers as it indicate "epsilon inside the square root", and i checked Pytorch doc, it use eps=1e-5, so i kept that. Thanks to your remind,  I find it's the same as i go deep into pytorch cpp code. The origin implementation is due to the typo of Pytorch doc which is fix already (https://github.com/huggingface/pytorch-transformers/pull/1089).
I'll switch that!

Reply:
BertLayerNorm switched to pytorch original LayerNorm.